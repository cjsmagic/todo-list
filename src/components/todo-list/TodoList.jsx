import PropTypes from "prop-types";
import { useState } from "react";
import TodoItem from "./TodoItem";

const propTypes = {
  list: PropTypes.array.isRequired,
  addTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  updateTodo: PropTypes.func.isRequired,
};

const TodoList = ({ list, addTodo, deleteTodo, updateTodo }) => {
  const [todo, setTodo] = useState("");
  const [tempText, setTempText] = useState(null);

  return (
    <div className="todo-list" id="124">
      <h1>Todo List</h1>
      <div className="todo-list__input-controls">
        <input
          type="text"
          placeholder="Enter todo"
          value={todo}
          onChange={(e) => setTodo(e.target.value)}
          data-testid="test-id-todo-input"
        />
        <button
          type="button"
          onClick={() => {
            addTodo({ name: todo, id: new Date().getTime() });
            setTodo("");
          }}
        >
          Add
        </button>
      </div>

      <div className="todo-list__items">
        {list.length === 0 && (
          <div className="todo-list__item todo-list__item--no-items">
            <div className="todo-list__item__name">No Items Found</div>
          </div>
        )}

        {list.map((item) => (
          <TodoItem
            key={item.id}
            item={item}
            tempText={tempText}
            setTempText={setTempText}
            updateTodo={updateTodo}
            deleteTodo={deleteTodo}
            items={list}
          />
        ))}
      </div>
    </div>
  );
};
TodoList.propTypes = propTypes;

export default TodoList;
