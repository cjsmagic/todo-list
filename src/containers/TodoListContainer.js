import PropTypes from "prop-types";
import React from 'react';
import { connect } from 'react-redux';
import TodoList from '../components/todo-list/TodoList';
import { ADD, DELETE, UPDATE } from '../redux/actions';

const props = {
    list: PropTypes.array.isRequired,
    addTodo: PropTypes.array.isRequired,
    deleteTodo: PropTypes.array.isRequired
};

const addTodo = (todo) => ({ type: ADD, payload: todo });

const deleteTodo = (todo) => ({ type: DELETE, payload: todo });

const updateTodo = (todo) => ({ type: UPDATE, payload: todo });

const TodoListContainer = ({ list, addTodo, deleteTodo }) => <TodoList
    list={list}
    addTodo={addTodo}
    deleteTodo={deleteTodo}
    updateTodo={updateTodo}
/>

const mapStateToProps = state => ({
    list: state.todoList
});

TodoListContainer.props = props;

export default connect(mapStateToProps, { addTodo, deleteTodo, updateTodo })(TodoListContainer)