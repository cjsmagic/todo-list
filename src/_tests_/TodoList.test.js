// __tests__/TodoList.test.js
import React from 'react'
import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor, screen, cleanup } from './testUtils'
import '@testing-library/jest-dom/extend-expect'
import TodoListContainer from '../containers/TodoListContainer';
import TodoList from '../components/todo-list/TodoList';


afterEach(() => {
    cleanup()
})

const deps = { initialState: { todoList: [] } };

const props = { list: [{ name: 'todo 1', id: new Date().getTime() }], addTodo: () => null, deleteTodo: () => null, updateTodo: () => null, }

test('it should load', async () => {
    render(<TodoListContainer />, deps);
    // screen.debug();
    expect(screen.getByRole('heading', { name: /todo list/i })).toHaveTextContent('Todo List')
})

it('renders according to design', () => {
    const tree = renderer
        .create(<TodoList {...props}
        />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});


test('it should add a new item', async () => {
    render(<TodoListContainer />, deps);
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).toBeTruthy())
    await waitFor(() => expect(inputElement.value).toBe(''))
    // screen.debug();
})


test('it should list items', async () => {
    const { container } = render(<TodoListContainer />, deps)
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))
    // screen.debug();
    const uniqueName2 = `${new Date().getTime()}__todo`;
    fireEvent.change(inputElement, { target: { value: uniqueName2 } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))


    await waitFor(() => {

        const items = Array.from(container.querySelectorAll('.todo-list__item'))
        expect(items.length).toEqual(2)
    })

    // screen.debug();
})



test('it should be deletable', async () => {
    render(<TodoListContainer />, deps)
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).not.toBeNull());
    // screen.debug();
    fireEvent.click(screen.getByRole('button', {
        name: /delete/i
    }));

    await waitFor(() => expect(screen.queryByText(uniqueName)).toBeNull());

    await waitFor(() => expect(screen.queryByText('No Items Found')).not.toBeNull());
    // screen.debug();
})


test('it should be editable', async () => {
    render(<TodoListContainer />, deps)
    // screen.debug();
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).not.toBeNull());
    // screen.debug();
    const editButton = screen.getByRole('button', {
        name: /edit/i
    });
    await waitFor(() => expect(editButton).not.toBeNull());
    fireEvent.click(editButton);
    // screen.debug();

    await waitFor(() => expect(editButton).not.toBeNull());
    fireEvent.click(editButton);

    const editElement = screen.getByTestId('test-id-todo-edit');
    const uniqueEditName = `${new Date().getTime()}__todo`;
    fireEvent.change(editElement, { target: { value: uniqueEditName } })
    // screen.debug();
    const updateButton = screen.getByRole('button', {
        name: /update/i
    });
    await waitFor(() => expect(updateButton).not.toBeNull());
    fireEvent.click(updateButton);

    await waitFor(() => expect(screen.getByText(uniqueEditName)).toBeTruthy())
    // screen.debug();
})