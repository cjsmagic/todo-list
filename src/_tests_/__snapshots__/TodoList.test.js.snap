// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders according to design 1`] = `
<div
  className="todo-list"
  id="124"
>
  <h1>
    Todo List
  </h1>
  <div
    className="todo-list__input-controls"
  >
    <input
      data-testid="test-id-todo-input"
      onChange={[Function]}
      placeholder="Enter todo"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
  <div
    className="todo-list__items"
  >
    <div
      className="todo-list__item"
    >
      <div
        className="todo-list__item__name"
      >
        todo 1
      </div>
      <button
        className="todo-list__item__edit todo-list__item__button"
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="todo-list__item__delete todo-list__item__button"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
